system/temperature.c

```diff
@@ -21,6 +21,141 @@
 
 #include "temperature.h"
 
+static uint16_t pci_read_pid(int bus, int dev, int func)
+{
+    return pci_config_read16(bus, dev, func, 2);
+}
+
+static uint16_t pci_read_vid(int bus, int dev, int func)
+{
+    return pci_config_read16(bus, dev, func, 0);
+}
+
+static int k8dev, k8fun;
+
+// find the first PCI device with the related IDs
+// On a multi-socketed K8, it grab only the first one since memtest only show
+// one temp
+static bool find_k8_temp(void)
+{
+    uint32_t pid, vid;
+
+    for (k8dev = 0; k8dev < 32; k8dev++) {
+        for (k8fun = 0; k8fun < 8; k8fun++) {
+            vid = pci_read_vid(0, k8dev, k8fun);
+            pid = pci_read_pid(0, k8dev, k8fun);
+            if (vid == PCI_VID_AMD && pid == 0x1103)
+                return true;
+        }
+    }
+    return false;
+}
+
+static int k8_temp(void)
+{
+    // k8temp is stored on a PCI device ([AMD] K8 [Athlon64/Opteron] Miscellaneous Control)
+#define K8_REG_TEMP   0xe4
+#define K8_SEL_CORE   0x04
+    uint32_t reg;
+    int temp = 0;
+
+    if (!find_k8_temp())
+       return 0;
+
+    if (cpuid_info.version.model == 4 && cpuid_info.version.stepping == 0)
+        return 0;
+    if (cpuid_info.version.model == 5 && cpuid_info.version.stepping <= 1)
+        return 0;
+    // TODO: backport Linux test is_rev_g_desktop() which add 21°
+
+    // We want temperature of core0
+    if (cpuid_info.version.model >= 0x40)
+        temp = K8_SEL_CORE;
+    pci_config_write8(0, k8dev, k8fun, K8_REG_TEMP, temp);
+    reg = pci_config_read32(0, k8dev, k8fun, K8_REG_TEMP);
+    // formula is taken from k8temp Linux driver
+    temp = (reg >> 16) & 0xFF;
+
+    // nobody will use memtest in a frozen env
+    if (temp < 49)
+        return 0;
+
+    return temp - 49;
+}
+
+// K10 temperature is stored in a PCI device [AMD] Family 10h Processor Miscellaneous Control
+// There is one PCI device per CPU socket
+static int k10dev, k10fun;
+
+#define CPUID_PKGTYPE_MASK      0xF0000000
+#define CPUID_PKGTYPE_F         0x00000000
+#define CPUID_PKGTYPE_AM2R2_AM3 0x10000000
+#define REG_DCT0_CONFIG_HIGH    0x094
+#define DDR3_MODE 1 << 8
+static bool has_erratum_319(void)
+{
+    uint32_t eax, ebx, ecx, edx;
+    uint32_t pkg_type, reg_dram_cfg;
+
+    if (cpuid_info.version.family != 0x10)
+        return false;
+
+    cpuid(0x80000001, 0, &eax, &ebx, &ecx, &edx);
```

Comment:
Not directly related to your patch, but I think that this is a good opportunity to store more extended feature flags in cpuid_feature_flags_t , as cpuid 80000001 is also needed for the AMD extended APIC space feature ( #70 et al.), and it's silly to call cpuid twice, then partially discard the output on both sides. Sam, Martin ?

Reply:
Yes, this should be done on cpuid.c

Reply:
Finally not so silly because the (useful) result of CPUID 0x80000001[EBX] is specific to K8, so it does make sense to run cpuid again later, only when this specific CPU has been detected.